### DISCUSSION TITLE
response: {"code":401,"message":"The request requires valid admin authorization token to be set.

### DISCUSSION THREAD

**Discussion by vincentgpower** [pocketbase/pocketbase#5558] [Q&A]
# curl -X GET http://localhost:8090/api/collections \
-H "Authorization: Admin eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MjgwNzQxNjgsImlkIjoiaGpjY3BzNGwyajg2ZmxiIiwidHlwZSI6ImFkbWluIn0.cXBSfaWljt8RxTs10OcnY5CJ6qiMnzlubbmb817gaew" \
-H "Content-Type: application/json"
> > **### {"code":401,"message":"The request requires valid admin authorization token to be set.","data":{}}**

ChatGPT said:

The 401 Unauthorized error with the message "The request requires valid admin authorization token to be set." suggests that the token is either invalid, expired, or not properly formatted when making the API request.

------> Trying same in GoLang 

I generate a token as an administrator and try use it, same token in the same code,  to add a collection. Its a JWT valid token and the date is +1 month.

go run  create_auth_token.go

Admin token Generated: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MjgwNjY5NTAsImlkIjoiaGpjY3BzNGwyajg2ZmxiIiwidHlwZSI6ImFkbWluIn0.rHQ1-xS0tgfy4TJPkq49o_QazqFf1fwHG22fOPupajQ

**Same Issue :**

Admin token Then used in same code: 
2024/09/20 18:35:50 Token validation failed: token validation failed: status code 401, response: {"code":401,"message":"The request requires valid admin authorization token to be set.","data":{}}
exit status 1

<img width="628" alt="Screenshot 2024-09-20 at 20 44 49" src="https://github.com/user-attachments/assets/647a5266-dd82-4285-87f7-db66bf7ae037">

The only clue may be - that the pocketbase sees my valid admin login as a guest ? 

Help !

------------->

The go code is below.

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
)

type AdminLoginResponse struct {
	Token string `json:"token"`
	Admin struct {
		ID    string `json:"id"`
		Email string `json:"email"`
	} `json:"admin"`
}

// Function to log in as an admin and retrieve the token
func adminLogin(email, password string) (string, error) {
	loginData := map[string]string{
		"identity": email,
		"password": password,
	}

	body, err := json.Marshal(loginData)
	if err != nil {
		return "", fmt.Errorf("failed to marshal login data: %v", err)
	}

	resp, err := http.Post("http://rooms.today:8090/api/admins/auth-with-password", "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", fmt.Errorf("failed to log in: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("login failed: status code %d", resp.StatusCode)
	}

	var result AdminLoginResponse
	err = json.NewDecoder(resp.Body).Decode(&result)
	if err != nil {
		return "", fmt.Errorf("failed to parse login response: %v", err)
	}

	return result.Token, nil
}

// Function to validate the token by making an authenticated request
func validateToken(token string) error {
	// URL of the endpoint to validate the token (fetch collections in this case)
	url := "http://rooms.today:8090/api/collections"

	// Create a new HTTP request
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %v", err)
	}

	// Set the Authorization header with the token
	req.Header.Set("Authorization", "Admin "+token)
	req.Header.Set("Content-Type", "application/json")

	// Send the request
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send request: %v", err)
	}
	defer resp.Body.Close()

	// Check if the request was successful
	if resp.StatusCode != http.StatusOK {
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Errorf("token validation failed: status code %d, response: %s", resp.StatusCode, string(body))
	}

	// Token is valid if we get a 200 OK
	fmt.Println("Token is valid!")
	return nil
}

func main() {
	// Replace with your admin email and password
	email := "posm@dol.cc"
	password := "961"

	// Step 1: Log in to get the admin token
	token, err := adminLogin(email, password)
	if err != nil {
		log.Fatalf("Failed to log in: %v", err)
	}

	fmt.Println("Admin token:", token)

	// Step 2: Validate the token by making an authenticated request
	err = validateToken(token)
	if err != nil {
		log.Fatalf("Token validation failed: %v", err)
	}
}

**Comment by LindezaGrey** [comment]
Shouldn't the header be

`req.Header.Set("Authorization", "Bearer "+token)
`

instead of

`req.Header.Set("Authorization", "Admin "+token)
`

**Reply by vincentgpower** [reply]
You are a hero, makes sense I could not find this in the documentation - Big Thanks


# vi listCollections.go
# go run listCollections.go
Admin token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MjgwODU4MDMsImlkIjoiaGpjY3BzNGwyajg2ZmxiIiwidHlwZSI6ImFkbWluIn0.av2bqggPFp2fTUMhtQOk1zUVHfHUON1-UACZ81NW5x8
Collections: {"page":1,"perPage":30,"totalItems":2,"totalPages":1,"items":[{"id":"_pb_users_auth_","created":"2024-09-20 17:55:39.054Z","updated":"2024-09-20 23:31:04.277Z","name":"users","type":"auth","system":false,"schema":[{"system":false,"id":"users_name","name":"name","type":"text","required":false,"unique":false,"options":{"min":null,"max":null,"pattern":""}},{"system":false,"id":"users_avatar","name":"avatar","type":"file","required":false,"unique":false,"options":{"maxSelect":1,"maxSize":5242880,"mimeTypes":["image/jpg","image/jpeg","image/png","image/svg+xml","image/gif"],"thumbs":null}},{"system":false,"id":"nszv3thz","name":"role","type":"select","required":false,"unique":false,"options":{"maxSelect":1,"values":["\"admin\"","\"editor\"","\"viewer\""]}}],"listRule":"id = @request.auth.id","viewRule":"id = @request.auth.id","createRule":"","updateRule":"id = @request.auth.id","deleteRule":"id = @request.auth.id","options":{"allowEmailAuth":true,"allowOAuth2Auth":true,"allowUsernameAuth":true,"exceptEmailDomains":null,"manageRule":null,"minPasswordLength":8,"onlyEmailDomains":null,"requireEmail":false}},{"id":"j7paebryw7jl8jy","created":"2024-09-20 23:34:34.865Z","updated":"2024-09-20 23:34:34.865Z","name":"user","type":"auth","system":false,"schema":[],"listRule":null,"viewRule":null,"createRule":null,"updateRule":null,"deleteRule":null,"options":{"allowEmailAuth":true,"allowOAuth2Auth":true,"allowUsernameAuth":true,"exceptEmailDomains":null,"manageRule":null,"minPasswordLength":8,"onlyEmailDomains":null,"requireEmail":false}}]}

#

**Reply by ganigeorgiev** [reply]
The `Bearer` prefix is optional and not needed. The format of the header is mentioned at the beginning in https://pocketbase.io/docs/authentication/.

Or in other words, simply `Authorization: TOKEN` should work.
